File: task/inplacereversal/ReverseLinkedList.java
package inplacereversal;

import fastandslowpointers.addTwoNumbers.ListNode;

public class ReverseLinkedList {
    public static void main(String[] args) {
        ListNode listNode = reverseList(
                new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))))
        );

        System.out.println(listNode);
    }

    public static ListNode reverseList(ListNode head) {
        if(head == null) return new ListNode();
        if(head.next == null) return head;

        ListNode prev = null, curr = head;

        while(curr != null) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }

        return prev;
    }
}

File: python/week2/ReverseLinkedList.py
from typing import Optional

from python.ListNode import ListNode


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        prev = None
        curr = head
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt

        return curr

if __name__ == '__main__':
    Solution().reverseList(ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5))))))
